### PR TITLE
Add greater than symbol to attribute escaping

### LINKED ISSUE #8779: HTML tags in image alt text causes formatting issues for the following block
When an image block has an alternative text with HTML markup (or anything resembling it) in it, the following block loses its styling.

Replication:
- Add an image block
- Add any alt text, with a HTML tag, `My<br>text`, but `My<awesome-tag>text` for example
- Add a paragraph block after it
- Add colors to the paragraph text

When viewing the page now, you will see that the image is there, but the paragraph block does not have its styles.

Images:

**Without markup:**
image

**With markup:**
image

There appears to be something funky going on with escaping values or something here. Sure markup isn't supported in alt texts, but users write strange things and we obviously need to get the escaping/sanitizing right.

### LINKED ISSUE #9915: Attribute escaping difference between Gutenberg and WP
**Describe the bug**
There appears to be a difference between how Gutenberg (specifically React) encodes HTML attributes, and how WordPress does, and this can lead to some mangling.

React encodes attributes according to the HTML spec, and `>` is allowed. This means `test > thing` is stored as-is in an attribute.

WordPress has other ideas, though. Here's an image created by Gutenberg. Note the `alt` tag has a value of `test > thing`  (I've added returns in examples to make them look better):

```html
<div class="wp-block-image">
<figure class="aligncenter">
<img src="http://latest.local/wp-content/uploads/2018/06/kirby.jpg" alt="test > thing" class="wp-image-10"/>
</figure>
</div>
```

WordPress renders this as:

```html
<div class="wp-block-image">
<figure class="aligncenter">
<img src="http://latest.local/wp-content/uploads/2018/06/kirby.jpg" alt="test > thing&#8221; class=&#8221;wp-image-10&#8243;/></figure></div>
```

Note how the terminating quote in the `alt` attribute has been smart-quoted, and then everything else falls to pieces.

The culprit is `wptexturize`, and specifically the `_get_wptexturize_split_regex` regex which 'parses' the HTML using regex. It sees the `>` in the attribute, thinks it's an HTML tag, and the fun begins.

To test:

```
wp shell
wptexturize('<div class="wp-block-image"><figure class="aligncenter"><img src="http://latest.local/wp-content/uploads/2018/06/kirby.jpg" alt="test > thing" class="wp-image-10"/></figure></div>’)

=>
string(197) "<div class="wp-block-image"><figure class="aligncenter">
<img src="http://latest.local/wp-content/uploads/2018/06/kirby.jpg" alt="test > thing&#8221; class=&#8221;wp-image-10&#8243;/></figure></div>"
```

Going further you can see the `>` causes the regex to split (array position 2 and 3):
```
wp shell
preg_split( _get_wptexturize_split_regex( '' ), '<div class="wp-block-image"><figure class="aligncenter"><img src="http://latest.local/wp-content/uploads/2018/06/kirby.jpg" alt="test > thing" class="wp-image-10"/></figure></div>', -1, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY )

=>
array(6) {
  [0] =>
  string(28) "<div class="wp-block-image">"
  [1] =>
  string(28) "<figure class="aligncenter">"
  [2] =>
  string(79) "<img src="http://latest.local/wp-content/uploads/2018/06/kirby.jpg" alt="test >"
  [3] =>
  string(29) " thing" class="wp-image-10"/>"
  [4] =>
  string(9) "</figure>"
  [5] =>
  string(6) "</div>"
}
```

Changing `wptexturize` seems a huge headache. Maybe the best action is for Gutenberg to encode all attributes in a way compatible with `_get_wptexturize_split_regex`, which probably means converting the `>` to `&gt;` (or just disabling `wptexturize` on every blog...):

```
wp shell
wptexturize('<div class="wp-block-image"><figure class="aligncenter">
<img src="http://latest.local/wp-content/uploads/2018/06/kirby.jpg" alt="test &gt; thing" class="wp-image-10"/></figure></div>')
=>
string(182) "<div class="wp-block-image"><figure class="aligncenter">
<img src="http://latest.local/wp-content/uploads/2018/06/kirby.jpg" alt="test &gt; thing" class="wp-image-10"/></figure></div>"
```

Note the above problem affects custom class names, and probably other blocks where data is directly output to an attribute.

This matches the problem described in #8779

### PR DESCRIPTION
**Pull request by johngodley**
Although `>` is a [valid attribute character](https://www.w3.org/TR/html5/syntax.html#elements-attributes), it is used by `wptexturize` in WordPress to split HTML tokens. Without escaping `wptexturize` will incorrectly tokenize a string, [causing problems](https://github.com/WordPress/gutenberg/issues/991). 

For example, add `test > thing` as the `alt` tag for an image in Gutenberg, and then view the post:

```html
<img src="kirby.jpg" alt="test > thing" class="wp-image-10"/>
```

Results in this HTML rendered in WordPress:

```html
<img src="kirby.jpg" alt="test > thing&#8221; class=&#8221;wp-image-10&#8243;/>
```

Encoding in Gutenberg prevents this problem while being transparent to the Gutenberg visual UI. It does have the effect that if you add a `>` into a block's attribute in HTML mode it will be replaced by `&gt;`, and this matches the same behaviour of `&` being replaced with `&amp;`.

The problem likely exists for other block types where an attribute is output directly based on user input. For example, custom class names.

Fixes #9915 
Fixes #8779

Note this won't fix any existing posts that have a `>` in an attribute. Re-saving these in Gutenberg should resolve the problem.

## How has this been tested?
Additional unit test added for the `>` symbol, and existing unit tests modified.

It can also be manually verified that adding `test > thing` to an image's `alt` tag is displayed as-is in the Gutenberg UI, and results in `test &gt; thing` appearing in the HTML.

```html
<img src="kirby.jpg" alt="test &gt; thing" class="wp-image-10"/>
```

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation.

### CODE REVIEW

**Review by jorgefilipecosta — APPROVED** [review]
This PR fixed the problem in my tests, and now it looks like we are doing the same the classic editor does 👍 
I think we can merge this fix for now, but I wonder if this is not a bug that should be fixed in wptexturize.

File: packages/element/src/serialize.js
Comment (aduth) [inline packages/element/src/serialize.js]: As far as `@wordpress/element` is concerned, it should have no knowledge of a `wptexturize` existing. Further, there's nothing about this module which implies any relationship to being used in the context of a post (i.e. it can be used to produce HTML which would be injected in an admin screen). These modules are meant to be independently useful without consideration being strongly bound to any one specific context. Holding ourselves honest to this guideline will prevent biasing implementation toward our specific needs.

Now, with that said, it may still be our best option moving forward, though I'm also wondering a few things:

- Can this be fixed from `wptexturize`?
- Can this be fixed by running a string replace on the result generated by a serialize call? Or do we need to be granular about the replacement occurring within an attribute value?
- And if so, should we consider extension points to our handling of various parts of the serialization? An attribute value _visitor_, for example?

Considering WordPress' needs as foreign can help surface where these sorts of extension points might be broadly useful.

Comment (johngodley) [reply packages/element/src/serialize.js]: > these modules are meant to be independently useful

I wasn't aware of that, but it's good to know. This did seem like a surely-this-is-too-easy kind of fix and I was expecting it to raise a question or two!

My first thought was to look at `wptexturize`. It's a monster of an old function and seems complicated to fix without triggering consequences somewhere. Saying that, I'm unfamiliar with the process, and don't know if Gutenberg is going to be tied to a particular base WP version etc. A fix there is certainly ideal, but tackling it in Gutenberg seems more immediately practical.

I then looked at fixing it inside the blocks themselves, but this places the burden on the block creator.

I looked at the serialize code. I did consider doing a string replace, but fixing a regex bug with another regex seemed like double trouble! It should be possible though, and we don't need to be granular about it - any unencoded `>` in an attribute will break the post. This could leave the core Gutenberg code to produce clean HTML, and we then have a 'texturizing' layer.

I also looked further into the serialization code and wondered if we could do it at a higher level - somehow traverse the block output and add the extra encoding to attributes. It got complicated, and I then came across `escapeAttribute`, which as solutions go seemed ideally magical.

By extension points do you mean hooks? That would work, and we could target different environments. If `wptexturize` is fixed in a future WordPress, the `>` encoding could be removed etc

Comment (aduth) [reply packages/element/src/serialize.js]: >By extension points do you mean hooks? That would work, and we could target different environments. If wptexturize is fixed in a future WordPress, the > encoding could be removed etc

In my mind I had pictured something like an AST / other tree visitor, which could be passed as an option to serialize, or in the creation of a custom serializer (e.g. expose serialize factory method to create / override default behavior).

Prior art:

- https://github.com/ctrlplusb/react-tree-walker#visitor
- https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md#toc-visitors

Comment (nosolosw) [reply packages/element/src/serialize.js]: @WordPress/gutenberg-core I know there is a core ticket for this https://core.trac.wordpress.org/ticket/45387 It doesn't seem like it has progressed much. Shall we land this fix as a preventive measure while the core fix is ongoing?

**Comment by johngodley** [comment]
Thanks for the nudge @nosolosw !

If the core ticket gets any movement then that may ultimately be the preferred route.

If not then based on Andrew's feedback it seems like a more comprehensive solution is preferred in the long run.

However, if this PR makes sense for the immediate future then I've rebased and fixed it up so it's working again. I can also create another ticket with the desired plans.

File: packages/escape-html/src/index.js
Comment (iseulde) [inline packages/escape-html/src/index.js]: Could use a comment explaining why it diverges from HTML spec.

Comment (johngodley) [reply packages/escape-html/src/index.js]: Ah yes, I think it got lost in the rebase. Added!

File: packages/escape-html/src/index.js
Comment (aduth) [inline packages/escape-html/src/index.js]: I think it would also be good to include an actionable note for the future maintainer. Something like:

```suggestion
 *
 * Note that if a resolution for Trac#45387 comes to fruition, it is no longer
 * necessary for `escapeGreaterThan` to exist in this module.
 *
 * See: https://core.trac.wordpress.org/ticket/45387
```

File: packages/escape-html/src/index.js
Comment (aduth) [inline packages/escape-html/src/index.js]: I think it would be best if we avoided exposing this on the public interface of the module, to avoid committing ourselves to future maintenance if it's arguably a workaround for a WordPress quirk.

Noting that it's exported for use by testing, alternatives could include one of either:

- Moving it to a separate file from which it is the default export, but is imported and then not subsequently exported by `index.js`
- Naming it as an [unstable API](https://github.com/WordPress/gutenberg/blob/83f03b18ee0cc29d3c309da2c5d9a39392eefa75/docs/contributors/coding-guidelines.md#experimental-apis) `__unstableEscapeGreaterThan`

**Review by aduth — DISMISSED** [review]

File: packages/escape-html/src/index.js
Comment (iseulde) [inline packages/escape-html/src/index.js]: Why are we even exposing it? If you're exposing it for tests, I'd recommend making a separate file and not export it in `index.js`.

Comment (aduth) [reply packages/escape-html/src/index.js]: Related: https://github.com/WordPress/gutenberg/pull/9963#discussion_r237872160

**Comment by mcsf** [comment]
Punting, as we seemingly haven't reached consensus yet.

**Comment by gziolo** [comment]
> Punting, as we seemingly haven't reached consensus yet.

What are next steps? Should we land it or close it?

**Comment by johngodley** [comment]
Ok, I’ve updated this so the unstable code is also in a separate file. I’ve left the tests in the same place as they seem to make more sense there. I’ve also included changes for the package version and history (pending a real date)

File: packages/escape-html/CHANGELOG.md
Comment (ellatrix) [inline packages/escape-html/CHANGELOG.md]: Is this the right format? I saw previous additions with `(Unreleased)`, unsure if this is matched anywhere. Cc @gziolo.

**Review by ellatrix — APPROVED** [review]

File: packages/escape-html/CHANGELOG.md
Comment (gziolo) [reply packages/escape-html/CHANGELOG.md]: `(Unreleased)` is the preferred one. At least it is documented this way:

https://github.com/WordPress/gutenberg/tree/master/packages#maintaining-changelogs

File: packages/element/src/serialize.js
Comment (mdawaffe) [reply packages/element/src/serialize.js]: > These modules are meant to be independently useful without consideration being strongly bound to any one specific context.

Using `>` in an attribute is valid HTML. Using `&gt;` in an attribute is also valid HTML. Important to this discussion: the two have the same meaning.

If we're aiming to be as general as possible, we should pick `&gt;` since it's useable in more contexts.

Comment (aduth) [reply packages/element/src/serialize.js]: > Using `>` in an attribute is valid HTML. Using `&gt;` in an attribute is also valid HTML. Important to this discussion: the two have the same meaning.

I don't disagree, but:

1. The expressed intention of this function was to apply the minimal transformations necessarily to transform an attribute value to be conformant as per the specification of what constitutes a [valid attribute value](https://w3c.github.io/html/syntax.html#elements-attributes).
2. The bug still remains in WordPress that [valid HTML becomes mangled upon save](https://core.trac.wordpress.org/ticket/45387) (whether from Gutenberg or elsewhere), but because there are few who feel comfortable touching the relevant texturize code, the impetus has been placed on integrators to implement workarounds, as we've done here.
3. None of this matters, as the pull request has been approved on the pragmatic consideration that this be solved.

Comment (mdawaffe) [reply packages/element/src/serialize.js]: You're right. Thank you for your patience and for humoring me :)

I wanted to use this package elsewhere and found I could not because of the details of its implementation. I attempting to understand why, I found this ticket. My comment was meant to be a springboard into that discussion elsewhere, but I instead derailed the conversation. My apologies.

Comment (aduth) [reply packages/element/src/serialize.js]: @mdawaffe Not at all derailed, certainly not. I think I'd stressed far too much the idea of being "generic" in my original comment, and it was a good prompt to reevaluate what I'd meant in terms of the stated purpose of a module / function.

**Comment by ellatrix** [comment]
Rebased this PR, but I have no idea why the tests are failing...

**Comment by westonruter** [comment]
Has the code here been reverted in any way? When adding `&gt;` or `&lt;` into an attribute value, the serializer forces it to be `>` and `<` respectively.

```html
<!-- wp:paragraph -->
<p><abbr title="1 > 2">False</abbr> and <abbr title="1 < 2">True</abbr></p>
<!-- /wp:paragraph -->
```

This gets rendered by WordPress as:

```html
<p><abbr title="1 > 2&#8243;>False</abbr> and <abbr title="1 < 2">True</abbr></p>
```

The “False and” text is erroneously consumed be the attribute:

image

In this case, it seems to be a problem with `wptexturize()`, as it is converting the attribute closing `"`  into `&#8243;`.

But if the block serializer output `>` as `&gt;` then this symptom could be avoided.

**Comment by aduth** [comment]
@westonruter Nothing more recent has occurred here to my knowledge. Could you create an issue which includes version information and sample code? I might wonder how your `abbr` elements are generated could have an impact on the serialization.

**Comment by westonruter** [comment]
@aduth cheers. Added here: https://github.com/WordPress/gutenberg/issues/15636.

**Pull request merged**
